File: app/src/main/java/com/example/mz/udacitypopularmovies/data/MovieProvider.java
package com.example.mz.udacitypopularmovies.data;

import android.content.ContentProvider;
import android.content.ContentValues;
import android.content.UriMatcher;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.net.Uri;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

/**
 * Created by mateusz.zak on 08.04.2017.
 */

public class MovieProvider extends ContentProvider {
    public static final int CODE_MOVIE = 100;
    public static final int CODE_FAVOURITE_MOVIES = 101;
    private static final int CODE_VIDEOS = 102;
    private static final int CODE_REVIEWS = 103;
    private static final UriMatcher sUriMatcher = buildUriMatcher();
    private MoviesDbHelper mOpenHelper;

    private static UriMatcher buildUriMatcher() {
        final UriMatcher matcher = new UriMatcher(UriMatcher.NO_MATCH);
        final String authority = MovieContract.CONTENT_AUTHORITY;

        matcher.addURI(authority, MovieContract.PATH_MOVIE + "/#", CODE_MOVIE);
        matcher.addURI(authority, MovieContract.PATH_MOVIE, CODE_FAVOURITE_MOVIES);
        matcher.addURI(authority, MovieContract.PATH_VIDEOS + "/#", CODE_VIDEOS);
        matcher.addURI(authority, MovieContract.PATH_REVIEWS + "/#", CODE_REVIEWS);
        return matcher;
    }

    @Override
    public boolean onCreate() {
        mOpenHelper = new MoviesDbHelper(getContext());
        return true;
    }

    @Override
    public int bulkInsert(@NonNull Uri uri, @NonNull ContentValues[] values) {
        final SQLiteDatabase db = mOpenHelper.getWritableDatabase();

        switch (sUriMatcher.match(uri)) {
            case CODE_MOVIE:
                db.beginTransaction();
                int rowsInserted = 0;
                try {
                    for (ContentValues value : values) {
                        long _id = db.insert(MovieContract.MovieEntry.TABLE_NAME, null, value);
                        if (_id != -1) {
                            rowsInserted++;
                        }
                    }
                    db.setTransactionSuccessful();
                } finally {
                    db.endTransaction();
                }
                if (rowsInserted > 0) {
                    getContext().getContentResolver().notifyChange(uri, null);
                }
                return rowsInserted;
            case CODE_FAVOURITE_MOVIES:
            case CODE_REVIEWS:
            case CODE_VIDEOS:
            default:
                return super.bulkInsert(uri, values);
        }
    }

    @Nullable
    @Override
    public Cursor query(@NonNull Uri uri, @Nullable String[] projection, @Nullable String selection, @Nullable String[] selectionArgs, @Nullable String sortOrder) {
        Cursor cursor;
        switch (sUriMatcher.match(uri)) {
            case CODE_MOVIE: {
                String movie_id = uri.getLastPathSegment();
                String[] selectionArguments = new String[]{movie_id};
                cursor = mOpenHelper.getReadableDatabase().query(MovieContract.MovieEntry.TABLE_NAME,
                        projection,
                        MovieContract.MovieEntry.COLUMN_MOVIE_ID + " = ? ",
                        selectionArguments,
                        null,
                        null,
                        sortOrder);
                break;
            }
            case CODE_FAVOURITE_MOVIES:
                cursor = mOpenHelper.getReadableDatabase().query(MovieContract.MovieEntry.TABLE_NAME,
                        projection,
                        selection,
                        selectionArgs,
                        null,
                        null,
                        sortOrder);
                break;
            case CODE_REVIEWS: {
                String movie_id = uri.getLastPathSegment();
                String[] selectionArguments = new String[]{movie_id};
                cursor = mOpenHelper.getReadableDatabase().query(MovieContract.ReviewEntry.TABLE_NAME,
                        projection,
                        MovieContract.MovieEntry.COLUMN_MOVIE_ID + " = ? ",
                        selectionArguments,
                        null,
                        null,
                        sortOrder);
                break;
            }
            case CODE_VIDEOS: {
                String movie_id = uri.getLastPathSegment();
                String[] selectionArguments = new String[]{movie_id};
                cursor = mOpenHelper.getReadableDatabase().query(MovieContract.VideoEntry.TABLE_NAME,
                        projection,
                        MovieContract.MovieEntry.COLUMN_MOVIE_ID + " = ? ",
                        selectionArguments,
                        null,
                        null,
                        sortOrder);
                break;
            }
            default:
                throw new UnsupportedOperationException("Unknown uri: " + uri);
        }
        cursor.setNotificationUri(getContext().getContentResolver(), uri);
        return cursor;
    }

    @Nullable
    @Override
    public String getType(@NonNull Uri uri) {
        throw new RuntimeException("getType is not implemented");
    }

    @Nullable
    @Override
    public Uri insert(@NonNull Uri uri, @Nullable ContentValues values) {
        return null;
    }

    @Override
    public int delete(@NonNull Uri uri, @Nullable String selection, @Nullable String[] selectionArgs) {
        return 0;
    }

    @Override
    public int update(@NonNull Uri uri, @Nullable ContentValues values, @Nullable String selection, @Nullable String[] selectionArgs) {
        return 0;
    }
}

File: app/src/main/java/com/mz/popmovies/data/MovieProvider.kt
package com.mz.popmovies.data

import android.content.ContentProvider
import android.content.ContentValues
import android.content.UriMatcher
import android.database.Cursor
import android.net.Uri
import timber.log.Timber

/**
 * Created by mateusz.zak on 08.04.2017.
 */
class MovieProvider : ContentProvider() {
    private var mOpenHelper: MoviesDbHelper? = null
    override fun onCreate(): Boolean {
        mOpenHelper = MoviesDbHelper(context)
        return true
    }

    override fun bulkInsert(uri: Uri, values: Array<ContentValues>): Int {
        val db = mOpenHelper!!.writableDatabase
        Timber.d("bulkInsert invoked")
        return when (sUriMatcher.match(uri)) {
            CODE_MOVIE, CODE_FAVOURITE_MOVIES -> {
                Timber.d("... for CODE_MOVIE")
                db.beginTransaction()
                var rowsInserted = 0
                try {
                    for (value in values) {
                        val _id = db.insert(MovieContract.MovieEntry.TABLE_NAME, null, value)
                        if (_id != -1L) {
                            rowsInserted++
                        }
                    }
                    db.setTransactionSuccessful()
                } finally {
                    db.endTransaction()
                }
                if (rowsInserted > 0) {
                    context!!.contentResolver.notifyChange(uri, null)
                }
                rowsInserted
            }
            CODE_REVIEWS, CODE_VIDEOS -> super.bulkInsert(uri, values)
            else -> super.bulkInsert(uri, values)
        }
    }

    override fun query(
        uri: Uri,
        projection: Array<String>?,
        selection: String?,
        selectionArgs: Array<String>?,
        sortOrder: String?
    ): Cursor? {
        val cursor: Cursor
        cursor = when (sUriMatcher.match(uri)) {
            CODE_MOVIE -> {

                //this is useful for checking if movie is in the favouries list
                val movie_id = uri.lastPathSegment
                val selectionArguments = arrayOf(movie_id)
                mOpenHelper!!.readableDatabase.query(
                    MovieContract.MovieEntry.TABLE_NAME,
                    projection,
                    MovieContract.MovieEntry.COLUMN_MOVIE_ID + " = ? ",
                    selectionArguments,
                    null,
                    null,
                    sortOrder
                )
            }
            CODE_FAVOURITE_MOVIES ->                 //this is used for displaying favourites
                mOpenHelper!!.readableDatabase.query(
                    MovieContract.MovieEntry.TABLE_NAME,
                    projection,
                    selection,
                    selectionArgs,
                    null,
                    null,
                    sortOrder
                )
            CODE_REVIEWS -> {
                val movie_id = uri.lastPathSegment
                val selectionArguments = arrayOf(movie_id)
                mOpenHelper!!.readableDatabase.query(
                    MovieContract.ReviewEntry.TABLE_NAME,
                    projection,
                    MovieContract.MovieEntry.COLUMN_MOVIE_ID + " = ? ",
                    selectionArguments,
                    null,
                    null,
                    sortOrder
                )
            }
            CODE_VIDEOS -> {
                val movie_id = uri.lastPathSegment
                val selectionArguments = arrayOf(movie_id)
                mOpenHelper!!.readableDatabase.query(
                    MovieContract.VideoEntry.TABLE_NAME,
                    projection,
                    MovieContract.MovieEntry.COLUMN_MOVIE_ID + " = ? ",
                    selectionArguments,
                    null,
                    null,
                    sortOrder
                )
            }
            else -> throw UnsupportedOperationException("Unknown uri: $uri")
        }
        cursor.setNotificationUri(context!!.contentResolver, uri)
        return cursor
    }

    override fun getType(uri: Uri): String? {
        throw RuntimeException("getType is not implemented")
    }

    override fun insert(uri: Uri, values: ContentValues?): Uri? {
        return null
    }

    override fun delete(uri: Uri, selection: String?, selectionArgs: Array<String>?): Int {
        when (sUriMatcher.match(uri)) {
            CODE_MOVIE -> {
                val movie_id = uri.lastPathSegment
                val selectionArguments = arrayOf(movie_id)
                return mOpenHelper!!.readableDatabase.delete(
                    MovieContract.MovieEntry.TABLE_NAME,
                    MovieContract.MovieEntry.COLUMN_MOVIE_ID + " = ? ",
                    selectionArguments
                )
            }
            CODE_FAVOURITE_MOVIES -> throw UnsupportedOperationException(
                "Unknown uri: $uri"
            )
        }
        return 0
    }

    override fun update(
        uri: Uri,
        values: ContentValues?,
        selection: String?,
        selectionArgs: Array<String>?
    ): Int {
        return 0
    }

    companion object {
        const val CODE_MOVIE = 100
        const val CODE_FAVOURITE_MOVIES = 101
        private const val CODE_VIDEOS = 102
        private const val CODE_REVIEWS = 103
        private val sUriMatcher = buildUriMatcher()
        private fun buildUriMatcher(): UriMatcher {
            val matcher = UriMatcher(UriMatcher.NO_MATCH)
            val authority = MovieContract.CONTENT_AUTHORITY
            matcher.addURI(authority, MovieContract.PATH_MOVIE + "/#", CODE_MOVIE)
            matcher.addURI(authority, MovieContract.PATH_MOVIE, CODE_FAVOURITE_MOVIES)
            matcher.addURI(authority, MovieContract.PATH_VIDEOS + "/#", CODE_VIDEOS)
            matcher.addURI(authority, MovieContract.PATH_REVIEWS + "/#", CODE_REVIEWS)
            return matcher
        }
    }
}